server/auth/serviceaccount/claims.go

```diff
@@ -17,32 +17,46 @@ func ClaimSetFor(restConfig *rest.Config) (*types.Claims, error) {
 	username := restConfig.Username
 	if username != "" {
 		return &types.Claims{Claims: jwt.Claims{Subject: username}}, nil
-	} else if restConfig.BearerToken != "" || restConfig.BearerTokenFile != "" {
-		bearerToken := restConfig.BearerToken
-		if bearerToken == "" {
-			// should only ever be used for service accounts
-			data, err := os.ReadFile(restConfig.BearerTokenFile)
-			if err != nil {
-				return nil, fmt.Errorf("failed to read bearer token file: %w", err)
-			}
-			bearerToken = string(data)
-		}
-		parts := strings.SplitN(bearerToken, ".", 3)
-		if len(parts) != 3 {
-			return nil, fmt.Errorf("expected bearer token to be a JWT and therefore have 3 dot-delimited parts")
-		}
-		payload := parts[1]
-		data, err := base64.RawStdEncoding.DecodeString(payload)
-		if err != nil {
-			return nil, fmt.Errorf("failed to decode bearer token's JWT payload: %w", err)
-		}
-		claims := &types.Claims{}
-		err = json.Unmarshal(data, &claims)
+	}
+	if restConfig.BearerToken == "" && restConfig.BearerTokenFile == "" {
+		return nil, nil
+	}
+
+	bearerToken := restConfig.BearerToken
+	if bearerToken == "" {
+		// should only ever be used for service accounts
+		data, err := os.ReadFile(restConfig.BearerTokenFile)
 		if err != nil {
-			return nil, fmt.Errorf("failed to unmarshal bearer token's JWT payload: %w", err)
+			return nil, fmt.Errorf("failed to read bearer token file: %w", err)
 		}
+		bearerToken = string(data)
+	}
+
+	parts := strings.SplitN(bearerToken, ".", 3)
+	if len(parts) != 3 {
+		return nil, fmt.Errorf("expected bearer token to be a JWT and therefore have 3 dot-delimited parts")
+	}
+	payload := parts[1]
+	data, err := base64.RawStdEncoding.DecodeString(payload)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode bearer token's JWT payload: %w", err)
+	}
+
+	claims := &types.Claims{}
+	err = json.Unmarshal(data, &claims)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal bearer token's JWT payload: %w", err)
+	}
+
+	// attempt to derive SA name and namespace from Subject
+	// "system:serviceaccount:argo:jenkins" -> "argo", "jenkins"
+	// note that the SA name can have a colon in it, although the rest cannot
+	parts = strings.SplitN(claims.Subject, ":", 4)
+	if len(parts) < 4 {
 		return claims, nil
-	} else {
-		return nil, nil
 	}
```

Comment:
So, is `claims.Subject` always set to that? I'm not really familiar, so f you have any link to share on that, it would be helpful, thanks.

Reply:
For a k8s SA token, yes. In the issue I experimented a bit with different types of k8s auth. An SA token from a `Secret` or from `kubectl create token` always has this type of subject. So primarily that was done via experimentation on actual tokens.

In this section of the k8s docs there's an example of a decoded token (this one is equivalent to one generated by `kubectl create token`): https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#serviceaccount-token-volume-projection

IIRC, all other forms of auth (e.g. kubeconfig) are not token auth, so they wouldn't reach here.